core/trino-main/src/main/java/io/trino/metadata/InternalNode.java

```diff
@@ -108,13 +109,16 @@ public boolean equals(Object obj)
             return false;
         }
         InternalNode o = (InternalNode) obj;
-        return nodeIdentifier.equals(o.nodeIdentifier);
+        return coordinator == o.coordinator &&
+                Objects.equals(nodeIdentifier, o.nodeIdentifier) &&
+                Objects.equals(internalUri, o.internalUri) &&
+                Objects.equals(nodeVersion, o.nodeVersion);
     }
 
     @Override
     public int hashCode()
     {
-        return nodeIdentifier.hashCode();
+        return Objects.hash(nodeIdentifier, internalUri, nodeVersion, coordinator);
```

Comment:
While it's important to include internalUri in the equals (see commit message for rationale), can this particular field be problematic for equals?

Reply:
cc @electrum you might know

Reply:
Arent' `internalUri, nodeVersion, coordinator` redundant to `nodeIdentifier`?

Reply:
yes, unless some aspects of the node changed, then no

Reply:
> yes, unless some aspects of the node changed, then no

I don't understand that. Seems like we could rely on `nodeIdentifier` to be unique. It seems if we cannot rely on that, then we might have a problem

Reply:
Depends whether we think about class-level design for InternalNode, the line here https://github.com/trinodb/trino/blob/c58d00132bf8dc5dfd615e8cc60caca07ef07e23/core/trino-main/src/main/java/io/trino/metadata/DiscoveryNodeManager.java#L281 which depends on equals being reasonable, or overall system design.

I agree that, if we have two InternalNode instances that have equal node identifier, but some other fields are different, then we are in a trouble (system-wise). Hiding this problem with a clever equals isn't doing any good, is it?